### DISCUSSION TITLE
Memory leak when opening and closing a new istance of a window often

### DISCUSSION THREAD

**Discussion by alessandro-gandolfi-ors** [AvaloniaUI/Avalonia#13167] [Q&A] [answered]
## Describe the bug
Opening and closing a window frequently increases usage of memory consistently, never decreasing. This is reproducible with both UsePlatformDetect() or UseHeadless(). The garbage collector doesn't free any memory even in longer periods of time. I added screenshots of the Program.cs code and the XAML of App and MainWindow. Glossario1View is a UserControl with some Grids, Textboxes, Textblocks, Borders (basically a styled table).

## To Reproduce
Put a breakpoint on the "return 0" of Main() (sample project attached) and see the memory rising in the diagnostic tool.
The Thread.Sleep was only used to better visualize the process memory "steps".

## Expected behavior
Memory decreases after the window is closed and not used anymore.

## Screenshots
![image](https://github.com/AvaloniaUI/Avalonia/assets/95684590/6d272a78-b454-4912-87ed-baa3a5e7b284)
![image](https://github.com/AvaloniaUI/Avalonia/assets/95684590/b934664d-5a32-43d3-95b2-e319d4c0a229)
![image](https://github.com/AvaloniaUI/Avalonia/assets/95684590/a2c38285-46a9-48be-aa2c-f33b1cda9af0)

## Environment
- OS: Windows 11 and Linux
- Avalonia-Version: 11.1.999-cibuild0040497-beta

## Additional context
Sample project
[AvaloniaMemLeak.zip](https://github.com/AvaloniaUI/Avalonia/files/12830128/AvaloniaMemLeak.zip)

**Comment by alessandro-gandolfi-ors** [comment]
Possibly related to #9662 and #12455

**Comment by alessandro-gandolfi-ors** [comment]
I'm not an expert about the garbage collector, I asked to a coworker to run the project in dotMemory since I don't have a license and it seems that most of the memory is gen 0.
![gen0_memory](https://github.com/AvaloniaUI/Avalonia/assets/95684590/56fbdee9-98a5-49f7-8ca0-b6610e6d6592)

Maybe related to #10111 but in my case changing the GC mode to low latency didn't change anything...

**Comment by maxkatz6** [comment]
@alessandro-gandolfi-ors have you tried to check what holds these objects? Must be something with headless backend, as non headless platform didn’t had an obvious memory leak on plains windows.

**Comment by thevortexcloud** [comment]
This may also be related to an issue I reported awhile back. I was never able to track down exactly what was causing it but it was fairly easy to reproduce and only occurred when switching views. I ended up just using a workaround to make sure the view never gets recreated. Which made the problem go away completely.

https://github.com/AvaloniaUI/Avalonia/issues/10968

**Comment by maxkatz6** [comment]
I don't think this pattern of headless usage is correct (nor documented). As there is no event loop in your code.

Checking in dotMemory, it confirms that window reference is kept in Dispatcher callbacks:
![image](https://github.com/AvaloniaUI/Avalonia/assets/3163374/3fb8678a-8f40-4d6f-9af2-88f7caf5eadd)


But since there is no event loop, there is nothing to run the dispatcher either.
Adding `Dispatcher.UIThread.RunJobs();` on each iteration removes "memory leak".

**Reply by maxkatz6** [reply]
Furthermore, if you want to have a proper headless platform running from the console app, without nunit or xunit, check this documentation page: https://docs.avaloniaui.net/docs/next/concepts/headless/headless-custom

HeadlessUnitTestSession will run a dedicated thread with an event loop, solving these issues.

**Reply by alessandro-gandolfi-ors** [reply]
Thanks for the support! However, trying to use `Dispatcher.UIThread.RunJobs();` doesn't change anything for the sample project I attached:
![image](https://github.com/AvaloniaUI/Avalonia/assets/95684590/6a753485-5825-4acc-9b52-d90400f34f53)
I tried adding it at the start of the loop, at the end of the loop, before and after the loop, the result is always the same, memory never decreses. I also tried giving it max priority, no changes.

Using `HeadlessUnitTestSession` would probably be the best option but I also need to use the Skia renderer, I followed [this documented pattern](https://docs.avaloniaui.net/docs/next/concepts/headless/#capturing-the-last-rendered-frame) for the sample project. Is it possible to enable the Skia rendered when using `HeadlessUnitTestSession.StartNew()`?

Another possibility would be to disable the creation of those jobs for the dispatcher if possible. I don't need to "run" any jobs, only show and close momentarily the windows.

I just want to be clear and say that I'm running tests for a bigger project, my company is searching and testing various possible solutions. We are still deciding the best option for us and we are aware this is an edge case for Avalonia, not fully supported at least. We are already in contact with the sales team, awaiting a response from them for more information about official support and understand if the project is doable using Avalonia or not. It seems for now there are few performance issues like the one I reported (but again, I'm also admitting is not the best use of it), otherwise it could work.

**Reply by maxkatz6** [reply, marked as answer]
> Thanks for the support! However, trying to use Dispatcher.UIThread.RunJobs(); doesn't change anything for the sample project I attached:

You might need to run jobs after Show as well and force Garbage Collector to run on each iteration. But still I don't think it's the best solution.

> Using HeadlessUnitTestSession would probably be the best option but I also need to use the Skia renderer

Instead of `HeadlessUnitTestSession.StartNew(typeof(App))` you can use `HeadlessUnitTestSession.StartNew(typeof(ClassWithBuildAvaloniaApp));` where `ClassWithBuildAvaloniaApp` type contains BuildAvaloniaApp method. Like, Program type in normal apps, but it can be anything.

**Reply by alessandro-gandolfi-ors** [reply]
It seems to have helped a bit! The memory usage went relatively down since I doubled the number of windows.
![image](https://github.com/AvaloniaUI/Avalonia/assets/95684590/103c81ac-4cda-45a6-a7b0-01f737bd7529)

I still have some doubts, initially the memory seemed stable, then it started going up again. The GC collected garbage once and the memory didn't go down, but at least it stopped increasing, while the spikes in CPU usage indicate that new windows were being created and closed. Do you consider this behavior normal? I have little to no knowledge to have an opinion about it.

**Reply by timunie** [reply]
GC is a complex topic in .NET in general. Not only related to Avalonia, but in .NET in general. You can read several articles on it. Basically, the GC happens if .NET thinks memory is too high or the App is idle. 

https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals

**Reply by maxkatz6** [reply]
@alessandro-gandolfi-ors well, since you create tons of objects, including native objects (from Skia) in a loop, and that .NET GC likes to keep memory allocated to the process, it might be as well normal. Or not. Depends.

I usually profile apps with dotTrace, where I can take a snapshot of memory and/or force GC, and it clears memory when I do so.

**Reply by alessandro-gandolfi-ors** [reply]
In the bigger project that does operations between `window.Open()` and `window.Close()` using headless sessions as suggested, garbage collections are executed but most if not all the gen0 memory is "converted" to unmanaged memory, never decreasing. I checked that everything that could be disposed by my code would be. At some point, after some minutes with longer tests, the memory keeps being taken in matter of a couple mbs with each new instance of a window. Garbage collections keep happening but the unmanaged memory never decreases, sometimes part of the gen0 is converted to gen1 instead of becoming all unmanaged. When I will have more time with dotMemory, I'm going to search better in the sections of the program... The same thing happens in the sample project, with a limited amount of windows so with fewer memory usage.
![high_unmanaged_mem](https://github.com/AvaloniaUI/Avalonia/assets/95684590/0e247879-f164-4919-8d47-2b60ecac6eae)

**Reply by alessandro-gandolfi-ors** [reply]
After more testing and having deployed the "bigger project" to a containerized environment with limited max memory, I can say that the problems I'm having with performance when running on my PC are probably not related to Avalonia, but how .NET and the OS manage memory. In the deployed project it doesn't grow undefinetly causing a crash and restart anymore, thanks to the support of @maxkatz6 (and @timunie on telegram). Using `HeadlessUnitTestSession.StartNew()`, even if it's not for unit tests that I'm using them for, seems to help freeing the memory after the windows are closed and the total memory keeps being stable instead of increasing. I'm selecting the comment to use the headless sessions with skia as the answer, as that is what I'm currently using. Thanks a lot again for the support (and sorry for the possible headaches I caused)!
